lib/fluent/plugin/in_tail.rb

```diff
@@ -537,86 +535,144 @@ def swap_state(pe)
         mpe = MemoryPositionEntry.new
         mpe.update(pe.read_inode, pe.read_pos)
         @pe = mpe
-        @io_handler.pe = mpe # Don't re-create IOHandler because IOHandler has an internal buffer.
-
         pe # This pe will be updated in on_rotate after TailWatcher is initialized
       end
 
       class StatWatcher < Coolio::StatWatcher
-        def initialize(path, log, &callback)
+        def initialize(watcher, &callback)
+          @watcher = watcher
           @callback = callback
-          @log = log
-          super(path)
+          super(watcher.path)
         end
 
         def on_change(prev, cur)
           @callback.call
         rescue
           # TODO log?
-          @log.error $!.to_s
-          @log.error_backtrace
+          @watcher.log.error $!.to_s
+          @watcher.log.error_backtrace
+        end
+      end
+
+
+      class FIFO
+        def initialize(from_encoding, encoding)
+          @from_encoding = from_encoding
+          @encoding = encoding
+          @buffer = ''.force_encoding(from_encoding)
+          @eol = "\n".encode(from_encoding).freeze
+        end
+
+        attr_reader :from_encoding, :encoding, :buffer
+
+        def <<(chunk)
```

Comment:
Uh, this is the weirdest part.  As `IO.readpartial(len, buf)` returns the same object as the second argument,  the encoding of the reused buffer (`@iobuf`) eventually becomes the same encoding as that specfied to `chunk.force_encoding()`.  Reverting back to the original encoding is necessary because for unknown reasons an assertion failure occurs for older versions of Ruby when a buffer with any double-byte or quad-byte encoding (UTF-16 / UCS-2 / UTF-32) is given to `readpartial()` as the second argument.

Try this:
```
a = open('/dev/urandom')
buf = ''.force_encoding(Encoding::UTF_16BE)
a.readpartial(2048, buf)
```